### PR TITLE
fix: Ignore older config lastmodified timestamps when fetching a config.

### PR DESCRIPTION
**Pull request by JamieSinn**
Also set clientdate on sdkconfig events, and regular tracked events.

### CODE REVIEW

File: configmanager.go [hunk 1]
```diff
@@ -236,9 +240,12 @@ func (e *EnvironmentConfigManager) fetchConfig(numRetriesRemaining int, minimumL
 	}
 	lastModifiedHeader := resp.Header.Get("Last-Modified")
 	if lastModifiedHeader != "" {
-		lastModifiedHeaderTS, parseError := time.Parse(time.RFC1123, lastModifiedHeader)
+		responseLastModified, parseError := time.Parse(time.RFC1123, lastModifiedHeader)
 		if parseError == nil {
-			if len(minimumLastModified) > 0 && lastModifiedHeaderTS.Before(minimumLastModified[0]) && numRetriesRemaining > 0 {
+			if storedLM, parseErr := time.Parse(time.RFC1123, e.localBucketing.GetLastModified()); parseErr == nil && storedLM.After(responseLastModified) {
```
Comment (suthar26) [inline configmanager.go:245]: this is the same as `lastModified` or the `minimumLastModified` would that not be a better comparison?

File: configmanager.go [hunk 2]
```diff
@@ -215,15 +215,19 @@ func (e *EnvironmentConfigManager) fetchConfig(numRetriesRemaining int, minimumL
...
-		lastModified = minimumLastModified[0].Format(time.RFC1123)
+	storedLM, err := time.Parse(time.RFC1123, lastModified)
+	if err != nil {
+		util.Warnf("Error parsing last modified time: %s\n", err)
 	}
-	if etag != "" {
-		req.Header.Set("If-None-Match", etag)
+	if len(minimumLastModified) > 0 && storedLM.Before(minimumLastModified[0]) {
+		lastModified = minimumLastModified[0].Format(time.RFC1123)
 	}
```
Comment (jonathannorris) [inline configmanager.go:224]: Can you explain what this is trying to do? where is `minimumLastModified` coming from?

File: configmanager.go [hunk 3]
```diff
@@ -236,9 +240,12 @@ func (e *EnvironmentConfigManager) fetchConfig(numRetriesRemaining int, minimumL
...
 	lastModifiedHeader := resp.Header.Get("Last-Modified")
 	if lastModifiedHeader != "" {
-		lastModifiedHeaderTS, parseError := time.Parse(time.RFC1123, lastModifiedHeader)
+		responseLastModified, parseError := time.Parse(time.RFC1123, lastModifiedHeader)
 		if parseError == nil {
-			if len(minimumLastModified) > 0 && lastModifiedHeaderTS.Before(minimumLastModified[0]) && numRetriesRemaining > 0 {
+			if storedLM.After(responseLastModified) {
+				return e.fetchConfig(numRetriesRemaining - 1)
```
Comment (jonathannorris) [inline configmanager.go:246]: should we add a warning here?

File: configmanager.go (hunk 2 above)
Comment (JamieSinn) [reply configmanager.go:224]: Check if MinimumLastModified is passed (length check) - if it exists; then check whether or not the stored last modified timestamp is before the passed new minimum. If it is - set that as the new header to set.

File: configmanager.go [hunk 4]
```diff
@@ -236,9 +240,12 @@ func (e *EnvironmentConfigManager) fetchConfig(numRetriesRemaining int, minimumL
...
+			}
+			if len(minimumLastModified) > 0 && responseLastModified.Before(minimumLastModified[0]) && numRetriesRemaining > 0 {
 				return e.fetchConfig(numRetriesRemaining-1, minimumLastModified...)
 			}
```
Comment (jonathannorris) [inline configmanager.go:250]: doesn't there need to be an else statement here that if its beyond the number of retries that it doesn't try and save the old request body?

File: configmanager.go (hunk 2 above)
Comment (JamieSinn) [reply configmanager.go:224]: minlastmodified is used by SSE to pass a minimum request timestamp that we should expect.

File: configmanager.go (hunk 3 above)
Comment (JamieSinn) [reply configmanager.go:246]: we could - but nothing really will be done from it - it's an inactionable warning

File: configmanager.go (hunk 4 above)
Comment (jonathannorris) [reply configmanager.go:250]: also need a test for this ^

Comment (JamieSinn) [reply configmanager.go:250]: I'll add an abort out otherways - but that's not where we would want to abort.

Comment (JamieSinn) [reply configmanager.go:250]: We do - it's the test I added

**Review by jonathannorris — APPROVED** [review]

**Pull request merged**